File: src/chicken/cli/commands/follow.clj
(ns chicken.cli.commands.follow
  (:require [chicken.repository :as rep]
            [clojure.string :refer (split blank?)]
            [chicken.cli.app :refer (pluginCommand trim-split)]))

(def ^:private TOKEN " follows ")

(defn validate [input]
  (let
    [[name followed] (trim-split input TOKEN)]
    (if (or (blank? name) (blank? followed))
      {:errors '("Invalid 'follows' format. Type 'Username follows Username'")}
      :ok)))

(defn follow [input]
  "If username and followed username from input both exist add followed username to the followed set of username"
  (let
    [[name followed] (trim-split input TOKEN)]
    (when (and (rep/existsUser? name) (rep/existsUser? followed))
      (rep/follow name followed))))

(pluginCommand :token TOKEN :do follow :validate validate)

File: src/main/scala/chicken/cli/commands/Follow.scala
package chicken.cli.commands

import chicken.cli.App

case class Follow() extends CliCommand {

  def token = " follows "

  override def validate(input: String) = {
    val trimSplit = App.trimSplit(input, token)
    if (trimSplit.length != 2 ||
      trimSplit(0) == null || trimSplit(0).isEmpty ||
      trimSplit(1) == null || trimSplit(1).isEmpty)
      "Invalid 'follows' format. Type 'Username follows Username'" :: Nil
    else
      Nil
  }

  def exec(input: String) = {
    val trimSplit = App.trimSplit(input, token)
    val name = trimSplit(0)
    val followed = trimSplit(1)
    if (App.rep.isUserExist(name) && App.rep.isUserExist(followed))
      App.rep.follow(name, followed)
  }

}
